spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
+             One can replace it e.g. with a parallel search.
+             *)
+            LET howManyToPop == IF topReq.isLeft THEN 2 ELSE 1 IN  
+            <<FALSE, SubSeq(requestStack, 1 + howManyToPop, Len(requestStack))>>
+        ELSE (*
+             Dichotomy: schedule search requests for the left and right branches
+             (and pop the top element off the stack).
+             In contrast to the English spec, these requests are not processed immediately,
+             but one-by-one in a depth-first order (mind the optimization above).
+             *)
+            LET rightReq == [isLeft |-> FALSE, startHeight |-> (lh + rh) \div 2, endHeight |-> rh]
+                leftReq ==  [isLeft |-> TRUE, startHeight |-> lh, endHeight |-> (lh + rh) \div 2]
+            IN
+            <<TRUE, <<leftReq, rightReq>> \o Tail(requestStack)>>
+
+OnResponseHeader ==
+  /\ state = "working"
+  /\ inEvent.type = "responseHeader"
+  /\ storedHeaders' = storedHeaders \union { inEvent.hdr }
+  /\ LET res == OneStepOfBisection(storedHeaders')
+         verdict == res[1]
+         newStack == res[2]
+     IN
+      /\ requestStack' = newStack
+      /\ IF newStack = << >>
+         THEN /\ outEvent' = [type |-> "finished", verdict |-> verdict]
+              /\ state' = "finished"
+         ELSE /\ outEvent' = RequestHeaderForTopRequest(newStack)
+              /\ state' = "working"  
+
+LCInit ==
+    /\ state = "init"
+    /\ outEvent = NoEvent
+    /\ requestStack = <<>>
+    /\ storedHeaders = {}
+
+LCNext ==
+  OnStart \/ OnResponseHeader
+            
+            
+(********************* Lite client + Environment + Blockchain *******************)
+Init == BC!Init /\ EnvInit /\ LCInit
 
 Next ==
-  (* the blockchain may progress by adding one more block,
-     provided that the ultimate height has not been reached yet *)
-  /\ \/ height < ULTIMATE_HEIGHT /\ AppendBlock
-     \/ height = ULTIMATE_HEIGHT /\ UNCHANGED <<height, blockchain>> 
-  (* as time is passing, the minimal trusted height may increase *)
-  /\ minTrustedHeight' \in minTrustedHeight..ULTIMATE_HEIGHT
+  \/ LCNext  /\ UNCHANGED bcvars /\ UNCHANGED envvars
+  \/ EnvNext /\ UNCHANGED bcvars /\ UNCHANGED lcvars
+  \/ BC!Next /\ UNCHANGED lcvars /\ UNCHANGED envvars
+
+
+(* The properties to check *)
+\* check this property to get an example of a terminating light client
+NeverStart == state /= "working"
+
+NeverFinish == state /= "finished"
+
+
+\* Correctness states that all the obtained headers are exactly like in the blockchain.
+\* This formula is equivalent to Accuracy in the English spec, that is, A => B iff ~A => ~B.
+CorrectnessInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = TRUE
+        => (\A hdr \in storedHeaders: hdr[1] = blockchain[hdr[1].height])
+
+PrecisionInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = FALSE
+        => (\E hdr \in storedHeaders: hdr[1] /= blockchain[hdr[1].height])
+        
```

Comment:
Can we define a property that light client (wrongly) trusted a header that is different from the one in the blockchain?

Reply:
We might want to modify `Lite Client Accuracy` property to never trust a header which is outside trusted period. More general, it would be great if these specs are standalone, i.e., that reader does not need to read English spec. Probably English spec should appear in form of comments here.

Reply:
`PositiveBeforeTrustedHeaderExpires` requires that whenever the lite client accepts the header, the trusted header is still within the trusting period. So far, TLC has not found a counterexample, but I believe it should find one.